sklearn/linear_model/tests/test_coordinate_descent.py

```diff
@@ -1484,20 +1481,22 @@ def test_multi_task_lasso_cv_dtype():
 
 @pytest.mark.parametrize("fit_intercept", [True, False])
 @pytest.mark.parametrize("alpha", [0.01])
-@pytest.mark.parametrize("normalize", [False, True])
 @pytest.mark.parametrize("precompute", [False, True])
-def test_enet_sample_weight_consistency(fit_intercept, alpha, normalize, precompute):
+@pytest.mark.parametrize("sparseX", [False, True])
+def test_enet_sample_weight_consistency(fit_intercept, alpha, precompute, sparseX):
     """Test that the impact of sample_weight is consistent."""
     rng = np.random.RandomState(0)
     n_samples, n_features = 10, 5
 
     X = rng.rand(n_samples, n_features)
     y = rng.rand(n_samples)
+    if sparseX:
+        X = sparse.csc_matrix(X)
     params = dict(
         alpha=alpha,
         fit_intercept=fit_intercept,
         precompute=precompute,
-        tol=1e-6,
+        tol=1e-9 if sparseX else 1e-6,
```

Comment:
this is surprising to me. What I see if that you apply the weights in the CD but I don't see why the dense and sparse cases would give a different numerical value after the same number of iterations. Is it also unexpected for you?

Reply:
Exactly. I suspect that I missed something in the cd solver's stopping criterion.

Reply:
Solved. I fixes 2 bugs of mine in the cd solver.